dex/wait/queue.go

```diff
@@ -0,0 +1,104 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package wait
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+const (
+	// These constants are used for a Waiter to inform the TickerQueue whether to
+	// run again or not.
+	TryAgain     = false
+	DontTryAgain = true
+)
+
+// Waiter is a function to run every recheckInterval until completion or
+// expiration. Completion is indicated when the TryFunc returns DontTryAgain.
+// Expiration occurs when TryAgain is returned after Expiration time.
+type Waiter struct {
+	// Expiration time is checked after the function returns TryAgain. If the
+	// current time > Expiration, ExpireFunc will be run and the waiter will be
+	// un-queued.
+	Expiration time.Time
+	// TryFunc is the function to run every recheckInterval until DontTryAgain is
+	// returned or Waiter expires.
+	TryFunc func() bool
+	// ExpireFunc is a function to run in the case that the Waiter expires.
+	ExpireFunc func()
+}
+
+// TickerQueue is a Waiter manager that checks a function periodically until
+// DontTryAgain is indicated.
+type TickerQueue struct {
+	waiterMtx       sync.RWMutex
+	waiters         []*Waiter
+	recheckInterval time.Duration
+}
+
+// NewTickerQueue is the constructor for a new TickerQueue.
+func NewTickerQueue(recheckInterval time.Duration) *TickerQueue {
+	return &TickerQueue{
+		recheckInterval: recheckInterval,
+		waiters:         make([]*Waiter, 0, 256),
+	}
+}
+
+// Wait attempts to run the (*Waiter).TryFunc until either 1) the function
+// returns the value DontTryAgain, or 2) the function's Expiration time has
+// passed. In the case of 2, the (*Waiter).ExpireFunc will be run.
+func (w *TickerQueue) Wait(settings *Waiter) {
+	if time.Now().After(settings.Expiration) {
+		log.Error("wait.TickerQueue: Waiter given expiration before present")
```

Comment:
This is the only place the package logs. I wonder if we need it.

Reply:
Would be nice not to need it.

Would you mind resolving the conflicts?  They look pretty trivial.